main/src/ca/uwaterloo/flix/language/phase/Kinder.scala

```diff
@@ -0,0 +1,1077 @@
+    case ResolvedAst.Annotation(name, exps0, loc) =>
+      mapN(Validation.traverse(exps0)(ascribeExpression(_, kinds, root))) {
+        exps => KindedAst.Annotation(name, exps, loc)
+      }
+  }
+
+  /**
+    * Performs kinding on the given expression under the given kind environment.
+    */
+  private def ascribeExpression(exp00: ResolvedAst.Expression, kenv: KindEnv, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindedAst.Expression, KindError] = exp00 match {
+    case ResolvedAst.Expression.Wild(loc) => KindedAst.Expression.Wild(Type.freshVar(Kind.Star), loc).toSuccess
+    case ResolvedAst.Expression.Var(sym, tpe0, loc) =>
+      mapN(ascribeType(tpe0, KindMatch.subKindOf(Kind.Star), kenv, root)) {
+        tpe => KindedAst.Expression.Var(sym, tpe, loc)
+      }
+    case ResolvedAst.Expression.Def(sym, loc) => KindedAst.Expression.Def(sym, Type.freshVar(Kind.Star), loc).toSuccess
+    case ResolvedAst.Expression.Sig(sym, loc) => KindedAst.Expression.Sig(sym, Type.freshVar(Kind.Star), loc).toSuccess
+    case ResolvedAst.Expression.Hole(sym, loc) => KindedAst.Expression.Hole(sym, Type.freshVar(Kind.Star), Type.freshVar(Kind.Bool), loc).toSuccess
+    case ResolvedAst.Expression.Unit(loc) => KindedAst.Expression.Unit(loc).toSuccess
+    case ResolvedAst.Expression.Null(loc) => KindedAst.Expression.Null(loc).toSuccess
+    case ResolvedAst.Expression.True(loc) => KindedAst.Expression.True(loc).toSuccess
+    case ResolvedAst.Expression.False(loc) => KindedAst.Expression.False(loc).toSuccess
+    case ResolvedAst.Expression.Char(lit, loc) => KindedAst.Expression.Char(lit, loc).toSuccess
+    case ResolvedAst.Expression.Float32(lit, loc) => KindedAst.Expression.Float32(lit, loc).toSuccess
+    case ResolvedAst.Expression.Float64(lit, loc) => KindedAst.Expression.Float64(lit, loc).toSuccess
+    case ResolvedAst.Expression.Int8(lit, loc) => KindedAst.Expression.Int8(lit, loc).toSuccess
+    case ResolvedAst.Expression.Int16(lit, loc) => KindedAst.Expression.Int16(lit, loc).toSuccess
+    case ResolvedAst.Expression.Int32(lit, loc) => KindedAst.Expression.Int32(lit, loc).toSuccess
+    case ResolvedAst.Expression.Int64(lit, loc) => KindedAst.Expression.Int64(lit, loc).toSuccess
+    case ResolvedAst.Expression.BigInt(lit, loc) => KindedAst.Expression.BigInt(lit, loc).toSuccess
+    case ResolvedAst.Expression.Str(lit, loc) => KindedAst.Expression.Str(lit, loc).toSuccess
+    case ResolvedAst.Expression.Default(loc) => KindedAst.Expression.Default(Type.freshVar(Kind.Star), loc).toSuccess
+    case ResolvedAst.Expression.Apply(exp0, exps0, loc) =>
+      for {
+        exp <- ascribeExpression(exp0, kenv, root)
+        exps <- Validation.traverse(exps0)(ascribeExpression(_, kenv, root))
+      } yield KindedAst.Expression.Apply(exp, exps, Type.freshVar(Kind.Star), Type.freshVar(Kind.Bool), loc)
+    case ResolvedAst.Expression.Lambda(fparam0, exp0, loc) =>
+      for {
+        fparam <- ascribeFormalParam(fparam0, kenv, root)
```

Comment:
I am ok with calling it `ascribeX`, but in most other phases we simply call the main recursive traversal for `visit`. Is it better to be precise?

Reply:
This is mostly an artifact from when more inference was going on: I had to distinguish between ascribing and inferring. I can change to `visit`.